### PR TITLE
Use smgrexists() instead of access() to enforce uniqueness of generated relfilenumber

### PR DESCRIPTION
**Pull request by knizhnik**
## Problem

Postgres is using `access()` function in `GetNewRelFileNumber` to check if assigned relfilenumber is not used for any other relation. This check will not work in Neon, because we do not have all files in local storage.

## Summary of changes

Use smgrexists() instead which will check at page server if such relfilenode is used.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by hlinnaka** [comment]
Thanks! Please also write test for this.

**Comment by knizhnik** [comment]
> Thanks! Please also write test for this.

I see only one way to reproduce the problem: force kids wraparound. It means 2^32 relation creations or `vacuum full`. 
But it will take unacceptable amount of time.

I can try to write some function with speed-up this overflow like `test_consume_xids`.
Any better idea?

File: pageserver/src/pgdatadir_mapping.rs [hunk 1]
```diff
@@ -284,6 +284,14 @@ impl Timeline {
...
+        let buf = version.get(self, DBDIR_KEY, ctx).await?;
+        let dbdirs = match DbDirectory::des(&buf).context("deserialization failure") {
+            Ok(dir) => Ok(dir.dbdirs),
+            Err(e) => Err(PageReconstructError::from(e)),
+        }?;
+        if !dbdirs.contains_key(&(tag.spcnode, tag.dbnode)) {
+            return Ok(false);
+        }
         // fetch directory listing
         let key = rel_dir_to_key(tag.spcnode, tag.dbnode);
```
Comment (tristan957) [inline pageserver/src/pgdatadir_mapping.rs:296]: You can move this line higher up to use it in your contains_key call.

File: pageserver/src/pgdatadir_mapping.rs [hunk 2]
```diff
@@ -284,6 +284,14 @@ impl Timeline {
...
             return Ok(true);
         }
+        let buf = version.get(self, DBDIR_KEY, ctx).await?;
+        let dbdirs = match DbDirectory::des(&buf).context("deserialization failure") {
+            Ok(dir) => Ok(dir.dbdirs),
+            Err(e) => Err(PageReconstructError::from(e)),
+        }?;
+        if !dbdirs.contains_key(&(tag.spcnode, tag.dbnode)) {
+            return Ok(false);
+        }
```
[the remark this answers is not in the document]
Comment (tristan957) [reply pageserver/src/pgdatadir_mapping.rs:294]: It looks like the first few lines of this block could be replaced by a call to `self.list_dbdirs()`.

File: pgxn/neon_test_utils/neontest.c [hunk 3]
```diff
@@ -72,8 +73,18 @@ _PG_init(void)
 #define neon_read_at_lsn neon_read_at_lsn_ptr
 
 /*
- * test_consume_xids(int4), for rapidly consuming XIDs, to test wraparound.
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply pgxn/neon_test_utils/neontest.c:75]: Sorry, really cut&pastte bug.
Fixed.

File: pageserver/src/pgdatadir_mapping.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (knizhnik) [reply pageserver/src/pgdatadir_mapping.rs:294]: > I'm trying to understand what the difference is between this block of code and the one underneath it. Could you explain that to me?

This block extracts DbDirectory (serialised hash map of all databases in cluster)
and underneath block extracts RelDirectory (serialised hash map of relations in the database).

The reason of this change is that in some cases compute can call smgrexists for database which is not yet initialized on PS.

[the remark this answers is not in the document]
Comment (knizhnik) [reply pageserver/src/pgdatadir_mapping.rs:294]: > It looks like the first few lines of this block could be replaced by a call to self.list_dbdirs().

Yeh, `list_dbdirs` is really very similar. But it doesn't take `version` as parameter. And we need to use use `version`, which performs lookup not only in KV storage but also in pending modifications.

File: pageserver/src/pgdatadir_mapping.rs (hunk 1 above)
Comment (knizhnik) [reply pageserver/src/pgdatadir_mapping.rs:296]: Sorry, but it is not possible. `dbdirs` hash map has key of type `(Oid, Oid)`
while `rel_dir_to_key` returns `Key`.

File: pageserver/src/pgdatadir_mapping.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tristan957) [reply pageserver/src/pgdatadir_mapping.rs:296]: Thanks!

**Review by tristan957 — APPROVED** [review]

**Review by skyzh — APPROVED** [review]

**Comment by ololobus** [comment]
@knizhnik all approvals are here, so I think it only waits for rebasing

**Pull request merged**